### PR TITLE
Restructure files and unified build system

### PR DESCRIPTION
**Pull request by samkhal**
I'll need to merge this into other pull requests so all the folder renaming works.

With this, running `make` at the top level will make all of the subcomponents. This includes automatic build of the LCM dependencies, and automatic build of the teensy-related things. Each subfolder that wants to be compiled should have a short makefile of it's own (see serial_bridge), and should be added to the list of nodes in the top-level makefile. (I'll probably make this second part automatic later).

---
This change is [Reviewable](https://reviewable.io/reviews/samkhal/biped/8)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+BUILD_DIR = build
+NODE_TARGET_NAME = main
+
+#path to this makefile, i.e. to the top level biped dir
+MKFILE_DIR := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
+
+LCM_DEFS = $(wildcard $(LCM_TYPES_DIR)/*.lcm)
+
+default: all
```
Comment (LoganTutt) [inline Makefile]: The different make rules should be put next to each other if possible to make it easier to read. They currently are interspersed between all the variable definitions. preferably all variable definitions should go at the beginning, followed by the rules.

**Review by LoganTutt — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (samkhal) [reply Makefile]: Can do, was just thinking it might be easier to read with the variables near the rules they are relevant for

Comment (LoganTutt) [reply Makefile]: Usually I put related variables near each other, with some separation between groupings. ie, all lcm variables near each other. I usually find that easiest to read and understand

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+export LCM_BDIR_ABS=$(MKFILE_DIR)$(LCM_B_DIR)
+
+LCM_HEADERS = $(patsubst $(LCM_TYPES_DIR)/%.lcm, $(LCM_B_DIR)/%.hpp, $(LCM_DEFS))
+$(LCM_B_DIR)/%.hpp: $(LCM_TYPES_DIR)/%.lcm
+	lcm-gen --cpp --cpp-std=c++11 --cpp-hpath=$(LCM_B_DIR) $^
+
+#----------- Build LCM Java: 
+LCM_JAVA_DIR = $(LCM_B_DIR)/$(LCM_PACKAGE_NAME)
+# Build .java and .class files
+$(LCM_JAVA_DIR)/%.class: $(LCM_TYPES_DIR)/%.lcm
```
Comment (LoganTutt) [inline Makefile:55]: I dont think this is called during `all`, is that intentional?

Comment (samkhal) [reply Makefile:55]: They're called by the JAR rule on line 50

Comment (LoganTutt) [reply Makefile:55]: In that case line 44 with the `mkdir` seems redundant as `LCM_B_DIR` must exist from the earlier `LCM_HEADERS` rule on line 37, correct?

File: Makefile (hunk 1 above)
Comment (samkhal) [reply Makefile]: yeah, makes sense. Cleaned things up a bit.

**Review by LoganTutt — APPROVED** [review]

File: Makefile (hunk 2 above)
Comment (samkhal) [reply Makefile:55]: Good call, you're right. Turns out lcm-gen makes the target directories anyway. Got rid of it and pushed it to master.

**Pull request merged**